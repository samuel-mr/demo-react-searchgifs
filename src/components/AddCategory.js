import React, { useState } from 'react'
import PropTypes from 'prop-types'

export const AddCategory = ({ setCategory }) => {
    const [value, setValue] = useState('')

    const onInputChange = (e) => {
        setValue(e.target.value)
    }

    const handleSubmit = (e) => {
        e.preventDefault()
        if (value.trim().length > 2) {
            setCategory(currentValue => [value, ...currentValue])
            setValue('')
        }
    }

    return (
        <form onSubmit={handleSubmit}>
            <p>{value}</p>
            <input type='text' value={value} onChange={onInputChange}></input>

        </ form >
    )

}

AddCategory.propTypes = {
    setCategory: PropTypes.func.isRequired
}