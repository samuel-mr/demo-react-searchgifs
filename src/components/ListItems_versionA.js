import React, { useState, useEffect } from 'react'
import { ListItemImage } from './ListItemImage'
import { getGifs } from '../helpers/getGifs'

export const Listitem = ({ category }) => {

    const [images, setImages] = useState([])

    useEffect(() => {
        getGifs(category)
            .then(imgs => setImages(imgs))
    }, [category])


    return (
        <>
            <h1>{category}</h1>
            <div className='card-grid'>
                {
                    images.map(item => (
                        <ListItemImage
                            key={item.id}
                            {...item}
                        />
                    ))
                }
            </div>
        </>
    )
}