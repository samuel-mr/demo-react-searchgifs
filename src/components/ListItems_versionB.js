import React from 'react'
import { ListItemImage } from './ListItemImage'
import { useFetchGifs } from '../hooks/useFetchGifs'

export const ListItems_versionB = ({ category }) => {

    const { data:images, loading } = useFetchGifs(category)


    return (
        <>
            <h1 className='animate__animated animate__slideInDown'>{category}</h1>
            {loading && <div className='animate__animated animate__flash'>Cargando... </div>}
            <div className='card-grid'>
                {
                    images.map(item => (
                        <ListItemImage
                            key={item.id}
                            {...item}
                        />
                    ))
                }
            </div>
        </>
    )
}