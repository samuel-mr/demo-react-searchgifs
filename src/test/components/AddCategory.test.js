import React from 'react'
import { shallow } from 'enzyme'
import { AddCategory } from '../../components/AddCategory'

//import '@testing-library/jest-dom'

describe('AddCategory Test', () => {

    // jest.fn es un wrapper para analizar un función
    const setCategories = jest.fn()
    let wrapper = shallow(<AddCategory setCategory={setCategories} />)

    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategory={setCategories} />)
    })

    test('El componente debe hacer match con su snapshot', () => {

        expect(wrapper).toMatchSnapshot()
    })

    test('El componente debe actualizar el valor cuando se cambia el texto del input', () => {
        const INPUT_INGRESADO = 'hola'
        const input = wrapper.find('input')

        // 1. simula la inserción de texto en el input
        input.simulate('change', { target: { value: INPUT_INGRESADO } })
        const parrafo = wrapper.find('p')
        // 2. verifica si el parrafo tiene ese valor ingreado
        expect(parrafo.text()).toBe(INPUT_INGRESADO)
    })

    test('No debe llamar a la funcion "setCategories" ya que no se ingresó ningun texto', () => {

        wrapper.find('form').simulate('submit', { preventDefault() { } })
        // 1. verifica que la función no halla sido llamada
        expect(setCategories).not.toHaveBeenCalled()
    })

    test('Debe llamar a la función "setCategories" y limpiar la caja de texto', () => {
        const INPUT_INGRESADO = 'nuevo texto'
        // 1. ingresa valores en el input
        wrapper.find('input').simulate('change', { target: { value: INPUT_INGRESADO } })
        // 2. envía los datos del formulario
        wrapper.find('form').simulate('submit', { preventDefault() { } })

        // 3. verifica que la función halla sido llamada 1 ves
        expect(setCategories).toHaveBeenCalledTimes(1)
        // 4. verifica que la función hayya recibido como argumento otra función
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function))

        const input_value = wrapper.find('input').prop('value')
        // 5. verifica que el input esté el blanco 
        expect(input_value).toBe('')
    })

})