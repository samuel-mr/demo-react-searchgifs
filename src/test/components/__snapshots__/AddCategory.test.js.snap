// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCategory Test El componente debe hacer match con su snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
